### PR TITLE
Implement nft_buy_offers and tests

### LINKED ISSUE #564: Implement nft_buy_offers in nextgen RPC system
Implement nextgen handler for `nft_buy_offers`.

### PR DESCRIPTION
**Pull request by godexsoft**
Fixes #564

### CODE REVIEW

File: src/rpc/handlers/NFTOffers.cpp [hunk 1]
```diff
@@ -33,7 +33,7 @@ namespace json = boost::json;
 
 namespace ripple {
 
-void
+inline void
```
Comment (cindyyan317) [inline src/rpc/handlers/NFTOffers.cpp:36]: does "inline" really work if it is used in cpp file?

**Comment by cindyyan317** [comment]
1 Can we put the common logic to base class ,then nft_sell_offers can reuse that part?
2 The marker logic seems not right for both old implementation and new implementation. If I understand correctly, the marker is the last element returning to the requester, the following response also contains this offer. I don't understand why offers size changes according to the marker status. WDYT?
 **ha i saw the logic, it pops the last one out before return**

File: src/rpc/ngHandlers/NFTBuyOffers.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    if (input.marker)
+    {
+        cursor = uint256(input.marker->c_str());
+
+        // We have a start point. Use limit - 1 from the result and use the very
+        // last one for the resume.
+        auto const sle =
+            [this, &cursor, &lgrInfo, &yield]() -> std::shared_ptr<SLE const> {
+            auto key = keylet::nftoffer(cursor).key;
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/NFTBuyOffers.cpp]: const?

File: src/rpc/ngHandlers/NFTBuyOffers.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        *sharedPtrBackend_,
+        directory,
+        cursor,
+        startHint,
+        lgrInfo.seq,
+        reserve,
+        {},
+        yield,
+        [&offers](ripple::SLE&& offer) {
+            if (offer.getType() == ripple::ltNFTOKEN_OFFER)
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/NFTBuyOffers.cpp:139]: the directory is only for the nft buy offers, do we still need to check the type here?
If other type is possible, we can not use size check to include marker field (line 151).

File: src/rpc/ngHandlers/NFTBuyOffers.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                return true;
+            }
+
+            return false;
+        });
+
+    if (auto status = std::get_if<Status>(&result))
+        return Error{*status};
+
+    if (offers.size() == reserve)
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/NFTBuyOffers.cpp:151]: If there are other types in this directory, the size might be less than reserve, but we still need to include the marker.

**Review by cindyyan317 — COMMENTED** [review]
Send back just to confirm if we need a baseclass for sell offer and buy offer

File: src/rpc/ngHandlers/NFTBuyOffers.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+    boost::json::object obj = {
+        {JS(nft_offer_index), to_string(offer.key())},
+        {JS(flags), offer[sfFlags]},
+        {JS(owner), toBase58(offer.getAccountID(sfOwner))},
+        {JS(amount), std::move(amount)},
+    };
+
+    if (offer.isFieldPresent(sfDestination))
+        obj.insert_or_assign(
```
Comment (cindyyan317) [inline src/rpc/ngHandlers/NFTBuyOffers.cpp:49]: other places use [], wonder why we use insert_or_assign here?

File: src/rpc/ngHandlers/NFTBuyOffers.cpp (hunk 3 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/NFTBuyOffers.cpp:139]: I have no idea - i just copy the previous implementation and make it testable :)

File: src/rpc/ngHandlers/NFTBuyOffers.cpp (hunk 4 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/NFTBuyOffers.cpp:151]: Good point. I copied the code tho so not sure if this is a bug or not. We can deal with it after all rpc is migrated and tested i think.

File: src/rpc/ngHandlers/NFTBuyOffers.cpp (hunk 5 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/NFTBuyOffers.cpp:49]: I think it's more efficient than [] either way and we probably should use it more. But as with other things - this code is just copied as is.

File: src/rpc/ngHandlers/NFTBuyOffers.cpp (hunk 2 above)
Comment (godexsoft) [reply src/rpc/ngHandlers/NFTBuyOffers.cpp]: yep, could be 👍

**Comment by godexsoft** [comment]
> Send back just to confirm if we need a baseclass for sell offer and buy offer

Yes or some other way for common logic. But i think it's better refactored in the next issue that covers `nft_sell_offers` rpc so i left it in one place for now.

**Comment by godexsoft** [comment]
> 1 Can we put the common logic to base class ,then nft_sell_offers can reuse that part? 2 The marker logic seems not right for both old implementation and new implementation. If I understand correctly, the marker is the last element returning to the requester, the following response also contains this offer. I don't understand why offers size changes according to the marker status. WDYT? **ha i saw the logic, it pops the last one out before return**

1 - yes, in the next issue :)
2 - idk. i'm not sure whether `nft*` marker/paging is tested by integration or not but there are some tests that refer to `nft_buy_offers` so i'd imagine this part is tested to some degree. eitherway i don't think we need to focus on this now while our goal is to make rpc testable and cover with tests as much as we can. let's leave fixing paging/marker till after migration is done completely, wdyt?

**Review by cindyyan317 — APPROVED** [review]
LGTM. Just checked the integration test, we have tested on marker and limit field.

**Pull request merged**